symengine/serialize-cereal.h

```diff
@@ -25,6 +25,113 @@
 namespace SymEngine
 {
 
+template <class Archive>
+class RCPBasicAwareOutputArchive : public Archive
+{
+    using Archive::Archive;
+
+public:
+    void save_rcp_basic(const RCP<const Basic> &ptr)
+    {
+        uintptr_t addr = (uintptr_t)(void *)ptr.get();
+        (*this)(addr);
+
+        auto id = _addresses.find(addr);
+        bool first_seen = (id == _addresses.end());
+        (*this)(first_seen);
+
+        if (not first_seen) {
+            return;
+        }
+        TypeID type_code = ptr->get_type_code();
+        save_typeid(*this, type_code);
+        switch (type_code) {
+#define SYMENGINE_ENUM(type, Class)                                            \
+    case type:                                                                 \
+        save_basic(*this, static_cast<const Class &>(*ptr));                   \
+        break;
+#include "symengine/type_codes.inc"
+#undef SYMENGINE_ENUM
+            default:
+                save_basic(*this, *ptr);
+        }
+        _addresses.insert(addr);
+    }
+
+private:
+    std::set<uintptr_t> _addresses;
+    //! Overload the rtti function to enable dynamic_cast
+    void rtti(){};
+};
+
+template <class Archive>
+class RCPBasicAwareInputArchive : public Archive
+{
+    using Archive::Archive;
+
+public:
+    template <class T>
+    RCP<const T> load_rcp_basic()
+    {
+        try {
+            uintptr_t addr;
+            (*this)(addr);
+
+            bool first_seen;
+            (*this)(first_seen);
+
+            if (not first_seen) {
```

Comment:
`SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /home/runner/work/symengine/symengine/symengine/serialize-cereal.h:83:21 in `

This is weird, it looks like a bool is saved in `save_rcp_basic` in the exact same way. I wonder if it would help to use uint8_t and perhaps use two magic numbers to represent true and false? (e.g. 42 and 137)